src/combinatorics/catalan.lean

```diff
@@ -0,0 +1,158 @@
+
+## TODO
+
+* Prove that the Catalan numbers enumerate many interesting objects.
+* Generalise to Catalan numbers associated to arbitrary Coxeter groups.
+
+-/
+
+open_locale big_operators
+open finset
+
+/-- The recursive definition of the sequence of Catalan numbers:
+`catalan (n + 1) = ∑ i : fin n.succ, catalan i * catalan (n - i)` -/
+def catalan : ℕ → ℕ
+| 0       := 1
+| (n + 1) := ∑ i : fin n.succ, have _ := i.2, have _ := nat.lt_succ_iff.mpr (n.sub_le i),
+             catalan i * catalan (n - i)
+
+@[simp] lemma catalan_zero : catalan 0 = 1 := by rw catalan
+
+lemma catalan_succ (n : ℕ) : catalan (n + 1) = ∑ i : fin n.succ, catalan i * catalan (n - i) :=
+by rw catalan
+
+/-- The explicit definition of the sequence of Catalan numbers via the central binomial
+coefficient.-/
+def catalan' (n : ℕ) :  ℚ :=
+1 / (n + 1) * nat.central_binom n
+
+@[simp] lemma catalan'_zero : catalan' 0 = 1 :=
+begin rw catalan', simp, end
+
+/-- A helper sequence that can be used to prove the equality of the recursive and the explicit
+definition using a telescoping sum argument. -/
+private def gosper_catalan (n j : ℕ) : ℚ :=
+nat.central_binom j * nat.central_binom (n - j) * (2 * j - n) / (2 * n * (n + 1))
+
+lemma gosper_trick (n i : ℕ) (h : i ≤ n) :
+gosper_catalan (n+1) (i+1) - gosper_catalan (n+1) i = catalan' i * catalan' (n - i) :=
+begin
+rw ← @mul_right_inj' _ _ (i + 1 : ℚ) _ _ (nat.cast_add_one_ne_zero i),
+rw ← @mul_right_inj' _ _ (n - i + 1 : ℚ) _ _ (by exact_mod_cast (n - i).succ_ne_zero),
+{ simp only [gosper_catalan, catalan'],
+  rw [mul_sub, mul_sub],
+  rw mul_div,
+  rw ← mul_assoc,
+  rw ← mul_assoc,
+  rw_mod_cast nat.succ_mul_central_binom_succ i,
+  rw mul_comm i.central_binom,
+  simp only [← mul_assoc],
+  rw_mod_cast mul_comm (n - i + 1),
+  nth_rewrite 1 mul_div,
+  push_cast,
+  simp only [← mul_assoc],
+  rw mul_assoc (i + 1 : ℚ),
+  simp_rw nat.sub_add_comm h,
+  rw_mod_cast nat.succ_mul_central_binom_succ,
+  simp_rw nat.add_sub_add_right,
+  rw mul_comm _ (n-i).central_binom,
+  rw mul_comm _ (n-i).central_binom,
+  rw ← mul_assoc (i+1),
+  rw mul_comm (i+1),
+  push_cast,
+  simp only [mul_div, ← mul_assoc],
+  rw_mod_cast mul_comm _ (n-i).central_binom,
+  simp only [← mul_div, ← mul_assoc],
+  push_cast,
+  simp only [mul_assoc],
+  rw ← mul_sub,
+  simp only [← mul_assoc],
+  rw mul_comm _ ((n-i).central_binom : ℚ),
+  congr,
+  rw mul_comm _ (i.central_binom : ℚ),
+  rw mul_comm _ (i.central_binom : ℚ),
+  rw mul_comm _ (i.central_binom : ℚ),
+  simp only [mul_assoc],
+  rw ← mul_sub,
+  congr,
+  ring_nf,
+  field_simp,
+  rw div_self,
+  { have h : ((n : ℚ) + (-(i : ℚ) + 1)) * (i + 1) = (i + 1) * n + (-i ^ 2 + 1),
+    { ring },
+    { rw [← h, div_self],
+      apply mul_ne_zero,
+      { rw ← add_assoc,
+        rw ← sub_eq_add_neg,
+        exact_mod_cast (n-i).succ_ne_zero, },
+      { exact_mod_cast i.succ_ne_zero } }, },
+  { exact_mod_cast nat.succ_ne_zero _ } },
```

Comment:
```suggestion
  have : (n:ℚ) + 1 ≠ 0 := by exact_mod_cast n.succ_ne_zero,
  have : (n:ℚ) + 1 + 1 ≠ 0 := by exact_mod_cast (n + 1).succ_ne_zero,
  have : (i:ℚ) + 1 ≠ 0 := by exact_mod_cast i.succ_ne_zero,
  have : (n:ℚ) - i + 1 ≠ 0 := by exact_mod_cast (n - i).succ_ne_zero,
  have h₁ : ((i:ℚ) + 1) * (i + 1).central_binom = 2 * (2 * i + 1) * i.central_binom,
  { exact_mod_cast nat.succ_mul_central_binom_succ i },
  have h₂ : ((n:ℚ) - i + 1) * (n - i + 1).central_binom
    = 2 * (2 * (n - i) + 1) * (n - i).central_binom,
  { exact_mod_cast nat.succ_mul_central_binom_succ (n - i) },
  simp only [gosper_catalan, catalan'],
  push_cast,
  field_simp,
  have H : n + 1 - i = n - i + 1 := sorry,
  rw H,
  linear_combination
    (2:ℚ) * (n - i).central_binom * (i + 1 - (n - i)) * (n + 1) * (n + 2) * ((n - i) + 1) * h₁
    - 2 * i.central_binom * (n + 1) * (n + 2) * (i - (n - i) - 1) * (i + 1) * h₂,
```

Reply:
Now that we have `linear_combination`, whenever I see a proof that looks like "algebraic manipulation in rings" I assume it should be easy :)